app/models/notification.rb

```diff
@@ -0,0 +1,20 @@
+class Notification < ApplicationRecord
+  belongs_to :notification_type
+  belongs_to :initiator, :class_name => User, :foreign_key => 'user_id'
+  belongs_to :subject, :polymorphic => true
+  belongs_to :cause, :polymorphic => true
+  has_many :notification_recipients, :dependent => :delete_all
+  has_many :recipients, :class_name => User, :through => :notification_recipients
+
+  def self.emit(type, opts = {})
+    type = NotificationType.find_by_name!(type)
+    notification = create(opts.merge!(:notification_type => type))
+    [notification, notification.create_recipients]
+  end
+
+  def create_recipients
+    subscribers = notification_type.subscriber_ids(subject, initiator)
+    notification_recipients.create! subscribers.collect { |id| {:user_id => id } }
```

Comment:
Recipients can be implemented as [nested attributes](http://api.rubyonrails.org/classes/ActiveRecord/NestedAttributes/ClassMethods.html)